cirq-core/cirq/contrib/bayesian_network/bayesian_network_gate.py

```diff
@@ -0,0 +1,198 @@
+        self._init_probs = init_probs
+        for _, params, cond_probs in arc_probs:
+            if not isinstance(params, tuple):
+                raise ValueError('Conditional prob params must be a tuple.')
+            if len(cond_probs) != 2 ** len(params):
+                raise ValueError('Incorrect number of conditional probs.')
+            for cond_prob in cond_probs:
+                if cond_prob < 0.0 or cond_prob > 1.0:
+                    raise ValueError('Conditional prob should be between 0 and 1.')
+        self._arc_probs = arc_probs
+
+    def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
+        parameter_names = [init_prob[0] for init_prob in self._init_probs]
+        qubit_map = dict(zip(parameter_names, qubits))
+
+        for param, init_prob in self._init_probs:
+            yield self._generate_got_set_for_init_prob(qubit_map[param], init_prob)
+
+        for target, params, cond_probs in self._arc_probs:
+            yield self._generate_gate_set_for_arc_prob(
+                qubit_map[target], [qubit_map[param] for param in params], cond_probs
+            )
+
+    @classmethod
+    def _generate_got_set_for_init_prob(cls, qubit, init_prob):
+        if init_prob is not None:
+            yield cirq.ry(cls._prob_to_angle(init_prob)).on(qubit)
+
+    @classmethod
+    def _generate_gate_set_for_arc_prob(cls, target, params, cond_probs):
+        # Here we deviate slightly from the original paper, by using Gray coding as described in:
+        # https://arxiv.org/abs/1306.3991
+        #
+        # The goal is to reduce the total number of gates, but the math is unchanged.
+
+        graycode = GrayCode(len(params))
+        previous_binary = '1' * (2 ** len(params))
+
+        for notted_binary in graycode.generate_gray():
+            # We get the NOT of the code because we want to start with 111...1 so that we don't
+            # need to have X gates on all qubits from the begining. This does not change the
+            # output and is simply an optimization.
+            binary = ''.join('1' if bit == '0' else '0' for bit in notted_binary)
+
+            # TODO(tonybruguier): Further reduce the number of gates in case the prob is 0.0. This
+            # would mean skipping a Gray code, so the accounting of the X gates must be done
+            # carefully.
+            for bit, previous_bit, param in zip(binary, previous_binary, params):
+                if bit != previous_bit:
+                    yield cirq.X(param)
+
+            yield cirq.ry(cls._prob_to_angle(cond_probs[int(binary, 2)])).on(target).controlled_by(
+                *params
+            )
+
+            previous_binary = binary
+
+        for previous_bit, param in zip(previous_binary, params):
+            if previous_bit == '0':
+                yield cirq.X(param)
+
+    @classmethod
+    def _prob_to_angle(cls, prob):
+        # From equation 13 of the paper.
+        return 2.0 * math.asin(math.sqrt(prob))
```

Comment:
Question: Paper says this should atan ?

Reply:
I hadn't noticed. I had read another source, but I think the two formula are identical:
![PXL_20220422_062151729 MP](https://user-images.githubusercontent.com/14946972/164616000-70a2bc06-0eab-4aef-9e34-919d9d109c09.jpg)

Note that I had not thought of this. I added a comment.